contracts/token/AssetRegistry.sol

```diff
@@ -0,0 +1,118 @@
+pragma solidity ^0.5.0;
+
+import "../interfaces/ISRC20Roles.sol";
+import "./SRC20.sol";
+import "../interfaces/IAssetRegistry.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+
+/**
+ * AssetRegistry holds the real-world/offchain properties of the various Assets being tokenized.
+ * It provides functions for getting/setting these properties.
+ */
+contract AssetRegistry is IAssetRegistry, Ownable {
+
+    struct AssetType {
+        bytes32 kyaHash;
+        string kyaUrl;
+        uint256 netAssetValueUSD;
+    }
+
+    address public _src20Factory;
+
+    mapping(address => AssetType) public assetList;
+
+    modifier onlyFactory() {
+        require(_src20Factory == msg.sender, "Caller not factory");
+        _;
+    }
+
+    modifier onlyDelegate(address src20) {
+        require(SRC20(src20)._roles().isDelegate(msg.sender), "Caller not delegate");
+        _;
+    }
+
+    constructor(address src20Factory) public {
+        _src20Factory = src20Factory;
+    }
+
+    /**
+     * Add an asset to the AssetRegistry
+     *
+     * @param src20 the token address.
+     * @param kyaHash SHA256 hash of KYA document.
+     * @param kyaUrl URL of token's KYA document (ipfs, http, etc.).
+     *               or address(0) if no rules should be checked on chain.
+     * @return True on success.
+     */
+    function addAsset(address src20, bytes32 kyaHash, string calldata kyaUrl, uint256 netAssetValueUSD)
+        external
+        onlyFactory
+        returns (bool)
+    {
+        require(assetList[src20].netAssetValueUSD == 0, 'Asset already added, try update functions');
```

Comment:
This `require` is not needed, as only `onlyFactory` can call this and it get called only once - on token creation.

Reply:
True, but Factory could be changed at some point making this function vulnerable. I think it makes sense that this function protects itself from wrong input, so I'll leave it in for now.